diffrax/solver/additive_srk.py

```diff
@@ -0,0 +1,284 @@
+from dataclasses import dataclass
+from typing import Optional, Tuple
+
+import equinox as eqx
+import jax
+import jax.lax as lax
+import jax.numpy as jnp
+import jax.tree_util as jtu
+import numpy as np
+from equinox.internal import ω
+
+from ..brownian.base import AbstractBrownianPath
+from ..custom_types import Array, Bool, DenseInfo, LevyVal, PyTree, Scalar
+from ..local_interpolation import LocalLinearInterpolation
+from ..solution import RESULTS
+from ..term import _ControlTerm, AbstractTerm, MultiTerm, ODETerm
+from .base import AbstractStratonovichSolver
+
+
+_ErrorEstimate = None
+_SolverState = None
+
+
+@dataclass(frozen=True)
+class StochasticButcherTableau:
+    """A Butcher Tableau for Additive-noise SRK methods."""
+
+    # Only supports explicit SRK so far
+    c: np.ndarray
+    b_sol: np.ndarray
+    b_error: Optional[np.ndarray]
+    a: list[np.ndarray]
+
+    # coefficients for W and H (of shape (len(c)+1,)
+    cW: np.ndarray
+    cH: Optional[np.ndarray]
+    bW: Scalar
+    bH: Optional[Scalar]
+
+    def __post_init__(self):
+        assert self.c.ndim == 1
+        for a_i in self.a:
+            assert a_i.ndim == 1
+        assert self.b_sol.ndim == 1
+        assert (self.b_error is None) or self.b_error.ndim == 1
+        assert self.c.shape[0] == len(self.a)
+        assert all(i + 1 == a_i.shape[0] for i, a_i in enumerate(self.a))
+        assert (self.b_error is None) or self.b_error.shape[0] == self.b_sol.shape[0]
+        assert self.c.shape[0] + 1 == self.b_sol.shape[0]
+        assert self.cW.shape[0] == self.b_sol.shape[0]
+        assert self.cH.shape[0] == self.b_sol.shape[0]
+        for i, (a_i, c_i) in enumerate(zip(self.a, self.c)):
+            assert np.allclose(sum(a_i), c_i)
+        assert np.allclose(sum(self.b_sol), 1.0)
+
+        # TODO: add checks for whether the method is FSAL
```

Comment:
ALIGN, for example is FSAL, and perhaps it could appear. But yes, you're right, FSAL is just not supported. This is a little remnent of when I thought I might implement it soon, but I realised it is not really relevant for the solvers that I currently have, so no point adding it. Anyway, I just deleted that.

But you'll be happy to know that I implemented embedded error estimates for both ShARK and SRA1, so adaptive time stepping works with both. And same for ALIGN.

Reply:
Great!